### PR TITLE
SLI-187 Collect anonymous statistics

### PR DESCRIPTION
**Pull request by dbmeneses**

### CODE REVIEW

File: src/main/java/org/sonarlint/intellij/telemetry/TelemetryClient.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    this.application = application;
+    this.projectManager = projectManager;
+  }
+
+  public void run(TelemetryStorage telemetryStorage) {
+    try {
+      String version = application.getVersion();
+      boolean connected = isAnyProjectConnected();
+      sendData(wsHelper, telemetryStorage, version, connected);
+    } catch (Exception e) {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/telemetry/TelemetryClient.java]: MAJOR Either log or rethrow this exception. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/org/sonarlint/intellij/telemetry/TelemetryClient.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      // fail silently
+      e.printStackTrace();
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/telemetry/TelemetryClient.java]: MINOR Use a logger to log this exception. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1148)

File: src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+  @Override public void disposeComponent() {
+    cancelScheduler();
+    saveData();
+  }
+
+  private void saveData() {
+    try {
+      storage.save(getStorageFilePath());
+    } catch (IOException e) {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java]: MAJOR Either log or rethrow this exception. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+  private boolean error = false;
+
+  public SonarLintTelemetry(SonarApplication application, ProjectManager projectManager) {
+    this.worker = new TelemetryClient(new WsHelperImpl(), application, projectManager);
+  }
+
+  private void loadStorage() {
+    Path filePath = getStorageFilePath();
+
+    if (Files.exists(filePath)) {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java]: MAJOR Replace this with a call to the "toFile().exists()" method [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3725)

File: src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    cancelScheduler();
+    saveData();
+  }
+
+  private void saveData() {
+    try {
+      storage.save(getStorageFilePath());
+    } catch (IOException e) {
+      // fail silently
+      e.printStackTrace();
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java]: MINOR Use a logger to log this exception. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1148)

File: src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    this.worker = new TelemetryClient(new WsHelperImpl(), application, projectManager);
+  }
+
+  private void loadStorage() {
+    Path filePath = getStorageFilePath();
+
+    if (Files.exists(filePath)) {
+      try {
+        storage = TelemetryStorage.load(filePath);
+      } catch (IOException e) {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java]: MAJOR Either log or rethrow this exception. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+  }
+
+  @Override public void initComponent() {
+    loadStorage();
+
+    try {
+      if (storage.enabled()) {
+        schedule();
+      }
+    } catch (Exception e) {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java]: MAJOR Either log or rethrow this exception. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+      // fail silently
+      e.printStackTrace();
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java]: MINOR Use a logger to log this exception. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1148)

File: src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+      } catch (IOException e) {
+        // ignore, we will recreate
+      }
+    }
+
+    if (storage == null) {
+      storage = new TelemetryStorage();
+      try {
+        storage.save(filePath);
+      } catch (IOException e) {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java]: MAJOR Either log or rethrow this exception. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1166)

File: src/test/java/org/sonarlint/intellij/telemetry/TelemetryClientTest.java [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+import static org.mockito.Mockito.when;
+
+public class TelemetryClientTest extends SonarTest {
+  private WsHelper wsHelper;
+  private ProjectManager projectManager;
+  private SonarApplication sonarApp;
+  private SonarLintProjectSettings projectSettings;
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
```
Comment (SonarTech) [inline src/test/java/org/sonarlint/intellij/telemetry/TelemetryClientTest.java]: MINOR Remove this unused "TemporaryFolder". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2924)

**Comment by SonarTech** [comment]
SonarQube analysis reported 11 issues
* ![MAJOR][MAJOR] 7 major
* ![MINOR][MINOR] 4 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [telemetry](https://github.com/SonarSource/sonarlint-intellij/blob/7807dbfca7f90d058ac93cdf1d2cbdb1a4c7fecc/src/main/java/org/sonarlint/intellij/telemetry): Add a 'package-info.java' file to document the 'telemetry' package [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1228)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+  private ScheduledFuture<?> scheduledFuture;
+
+  public SonarLintTelemetry(SonarApplication application, ProjectManager projectManager) {
+    this.worker = new TelemetryClient(new WsHelperImpl(), application, projectManager);
+  }
+
+  private boolean loadStorage() {
+    Path filePath = getStorageFilePath();
+
+    if (Files.exists(filePath)) {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java]: MAJOR Replace this with a call to the "toFile().exists()" method [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3725)

File: src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    storage.setEnabled(enabled);
+    saveData();
+  }
+
+  @Override public void initComponent() {
+    try {
+      if (loadStorage()) {
+        scheduleUpload();
+      }
+    } catch (Exception e) {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java]: MAJOR Either log or rethrow this exception. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1166)

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues
* ![MAJOR][MAJOR] 6 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+
+  public SonarLintTelemetry(SonarApplication application, ProjectManager projectManager) {
+    this.worker = new TelemetryClient(new WsHelperImpl(), application, projectManager);
+  }
+
+  private boolean loadStorage() {
+    Path filePath = getStorageFilePath();
+    storage = null;
+
+    if (Files.exists(filePath)) {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java]: MAJOR Replace this with a call to the "toFile().exists()" method [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3725)

File: src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+      try {
+        storage = TelemetryStorage.load(filePath);
+      } catch (Exception e) {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java]: MAJOR Either log or rethrow this exception. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/org/sonarlint/intellij/telemetry/TelemetryStorage.java [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+        }
+
+        long maxUseDays = telemetry.installDate().until(telemetry.lastUseDate(), DAYS) + 1;
+        if (telemetry.numUseDays() > maxUseDays) {
+          telemetry.setNumUseDays(maxUseDays);
+          telemetry.setLastUseDate(today);
+        }
+      }
+    } else {
+      return new TelemetryStorage();
```
Comment (janos-ss) [inline src/main/java/org/sonarlint/intellij/telemetry/TelemetryStorage.java]: I'd flip the if-else branches to make this an early return and the rest of the method flatter

File: src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+import org.sonarsource.sonarlint.core.telemetry.Telemetry;
+
+public class SonarLintTelemetry implements ApplicationComponent {
+  private static final String PRODUCT = "SonarLint IntelliJ";
+  private static final String STORAGE_FILENAME = "sonarlint_usage";
+
+  private final SonarApplication application;
+  private final ProjectManager projectManager;
+
+  Telemetry telemetry;
```
Comment (janos-ss) [inline src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java]: Can be private?

File: src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+  private boolean isAnyProjectConnected() {
+    Project[] openProjects = projectManager.getOpenProjects();
+
+    for (Project p : openProjects) {
+      SonarLintProjectSettings projectSettings = SonarLintUtils.get(p, SonarLintProjectSettings.class);
+      if (projectSettings.isBindingEnabled()) {
+        return true;
+      }
+    }
+    return false;
```
Comment (janos-ss) [inline src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java]: Would it be easier to read if rewritten with a stream?

File: src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+      telemetry.enable(enabled);
+    }
+  }
+
+  @Override public void initComponent() {
+    try {
+      this.telemetry = new Telemetry(getStorageFilePath());
+      this.scheduledFuture = JobScheduler.getScheduler().scheduleWithFixedDelay(this::upload,
+        1, TimeUnit.HOURS.toMinutes(6), TimeUnit.MINUTES);
+    } catch (Exception e) {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java]: MAJOR Either log or rethrow this exception. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+  @Override public void disposeComponent() {
+    if (scheduledFuture != null) {
+      scheduledFuture.cancel(false);
+      scheduledFuture = null;
+    }
+    try {
+      if (telemetry != null) {
+        telemetry.save();
+      }
+    } catch (IOException e) {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/telemetry/SonarLintTelemetry.java]: MAJOR Either log or rethrow this exception. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1166)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

**Pull request merged**